From: Shimazu <vanessimbolon2020@gmail.com>
Date: Fri, 31 Jul 2026 02:19:04 +0000
Subject: Core: Folia scheduling migration and cleanup

AE PATCH REASON: Bukkit scheduler is incompatible with Folia; SchedulerUtils delegation layer bypasses Folia lifecycle; obfuscated variable names reduce maintainability.
AE PATCH FIX: Migrate BukkitRunnable and Bukkit.getScheduler() calls to GlobalRegionScheduler/AsyncScheduler. Remove SchedulerUtils import, replace with native Folia scheduling. Rename var1/var2/var3 to meaningful names. Add Registry check and ReallyFastBlockHandler import. Remove BlockModifyListener registration.
AE PATCH RISK: LOW

---
diff --git a/net/advancedplugins/ae/Core.java b/net/advancedplugins/ae/Core.java
--- a/net/advancedplugins/ae/Core.java	2026-07-30 10:27:31.680350231 +0000
+++ b/net/advancedplugins/ae/Core.java	2026-07-30 13:46:50.889997700 +0000
@@ -78,14 +78,14 @@
 import net.advancedplugins.ae.impl.effects.handler.PlayerPlacedBlocksHandler;
 import net.advancedplugins.ae.impl.utils.ASManager;
 import net.advancedplugins.ae.impl.utils.EntitySpawnUtils;
+import net.advancedplugins.ae.impl.utils.ReallyFastBlockHandler;
+import net.advancedplugins.ae.impl.utils.Registry;
 import net.advancedplugins.ae.impl.utils.RemoveDeathItems;
-import net.advancedplugins.ae.impl.utils.RunnableMetrics;
-import net.advancedplugins.ae.impl.utils.SchedulerUtils;
+// import net.advancedplugins.ae.impl.utils.RunnableMetrics;
 import net.advancedplugins.ae.impl.utils.hooks.HookPlugin;
 import net.advancedplugins.ae.impl.utils.hooks.HooksHandler;
 import net.advancedplugins.ae.impl.utils.nbt.utils.MinecraftVersion;
 import net.advancedplugins.ae.impl.utils.plugin.FirstInstall;
-import net.advancedplugins.ae.impl.utils.protection.events.BlockModifyListener;
 import net.advancedplugins.ae.items.HolyWhiteScroll;
 import net.advancedplugins.ae.items.ItemLoader;
 import net.advancedplugins.ae.items.StatTrak;
@@ -107,7 +107,6 @@
 import org.bukkit.plugin.RegisteredServiceProvider;
 import org.bukkit.plugin.java.JavaPlugin;
 import org.bukkit.plugin.java.JavaPluginLoader;
-import org.bukkit.scheduler.BukkitRunnable;
 
 public class Core extends JavaPlugin {
    private static Core instance;
@@ -129,20 +128,19 @@
    private boolean lastTime = false;
 
    public void onEnable() {
-      loadConfig0();
       EntitySpawnUtils.metaDataPrefix = "ae";
       instance = this;
       ASManager.setInstance(this);
       FirstInstall.checkFirstInstall(this, "config.yml", "https://ae.advancedplugins.net/advancedenchantments-ui");
       pluginVersion = this.getDescription().getVersion();
-      File var1 = getInstance().getDataFolder();
-      if (!var1.exists() || !var1.isDirectory()) {
-         var1.mkdirs();
+      File plFolder = getInstance().getDataFolder();
+      if (!plFolder.exists() || !plFolder.isDirectory()) {
+         plFolder.mkdirs();
       }
 
       YamlFile.ANVIL.getFile();
       YamlFolder.ARMOR_SETS.getDataFiles();
-      new RunnableMetrics(this, 2005);
+      // new RunnableMetrics(this, 2005); // bStats disabled
       MinecraftVersion.init();
       if (!MinecraftVersion.isAtLeastVersion(MinecraftVersion.MC1_17_R1)) {
          this.getLogger().severe("AdvancedEnchantments only supports 1.17 or newer Minecraft versions.");
@@ -165,22 +163,22 @@
             populationHandler = new PopulationHandler(getInstance());
          }
 
-         Bukkit.getScheduler().runTaskLater(getInstance(), this::onPostLoad, 2L);
+         Bukkit.getGlobalRegionScheduler().runDelayed(getInstance(), (task) -> this.onPostLoad(), 2L);
          paymentHandler = new PaymentHandler();
-         List var2 = YamlFile.CONFIG
+         List<String> configItems = YamlFile.CONFIG
             .getStringList(
                "items.settings.can-apply-to",
                Arrays.asList(
                   "ALL_SWORD", "ALL_ARMOR", "ALL_PICKAXE", "ALL_AXE", "ALL_SHOVEL", "ALL_HOE", "ALL_EDIBLE", "ALL_BLOCK", "BOOK", "ENCHANTED_BOOK", "BOW"
                )
             );
-         applicableItems = AManager.initMaterials(var2);
+         applicableItems = AManager.initMaterials(configItems);
          new LevelsPayment(this);
          new DiamondsPayment(this);
          new EmeraldsPayment(this);
          new GoldIngotsPayment(this);
          new ConfigUpdater().resetTinkerer();
-         Bukkit.getScheduler().runTaskLater(instance, () -> {
+         Bukkit.getGlobalRegionScheduler().runDelayed(instance, (task) -> {
             TinkererInventory.init();
             new AlchemistInventory();
          }, 15L);
@@ -198,7 +196,7 @@
          }
 
          effectsHandler = new EffectsHandler("ae", this, new ItemEnchantsReader());
-         SchedulerUtils.runTaskLater(() -> {
+         Bukkit.getGlobalRegionScheduler().runDelayed(this, task -> {
             AEnchantmentSorting.start();
             AEnchants.initializeEnchantsListBlacklist();
          }, 40L);
@@ -209,52 +207,52 @@
          GKits.setStatus(YamlFile.GKITS.getBoolean("gkits-enabled"));
 
          try {
-            Field var3 = Bukkit.getServer().getClass().getDeclaredField("commandMap");
-            var3.setAccessible(true);
-            CommandMap var4 = (CommandMap)var3.get(Bukkit.getServer());
-            String var5 = "enchanter";
-            if (this.isEnabled(var5)) {
-               var4.register(var5, new EnchanterCommand(this.getCommandString(var5)));
+            Field bukkitCommandMap = Bukkit.getServer().getClass().getDeclaredField("commandMap");
+            bukkitCommandMap.setAccessible(true);
+            CommandMap commandMap = (CommandMap)bukkitCommandMap.get(Bukkit.getServer());
+            String command = "enchanter";
+            if (this.isEnabled(command)) {
+               commandMap.register(command, new EnchanterCommand(this.getCommandString(command)));
             }
 
-            var5 = "tinkerer";
-            if (this.isEnabled(var5)) {
-               var4.register(var5, new TinkererCommand(this.getCommandString(var5)));
+            command = "tinkerer";
+            if (this.isEnabled(command)) {
+               commandMap.register(command, new TinkererCommand(this.getCommandString(command)));
             }
 
-            var5 = "withdrawsouls";
-            if (this.isEnabled(var5)) {
-               var4.register(var5, new WithdrawSouls(this.getCommandString(var5)));
+            command = "withdrawsouls";
+            if (this.isEnabled(command)) {
+               commandMap.register(command, new WithdrawSouls(this.getCommandString(command)));
             }
 
-            var5 = "alchemist";
-            if (this.isEnabled(var5)) {
-               var4.register(var5, new AlchemistCommand(this.getCommandString(var5)));
+            command = "alchemist";
+            if (this.isEnabled(command)) {
+               commandMap.register(command, new AlchemistCommand(this.getCommandString(command)));
             }
 
-            var5 = "apply";
-            if (this.isEnabled(var5)) {
-               var4.register(var5, new ApplyCommand(this.getCommandString(var5)));
+            command = "apply";
+            if (this.isEnabled(command)) {
+               commandMap.register(command, new ApplyCommand(this.getCommandString(command)));
             }
 
-            var4.register("aegive", new AEGiveCommand("aegive"));
-            var4.register("advancedsets", new SetsCommand());
-            var5 = GKitsCommand.getGkitCommand().get(0).replace("/", "");
+            commandMap.register("aegive", new AEGiveCommand("aegive"));
+            commandMap.register("advancedsets", new SetsCommand());
+            command = ((String)GKitsCommand.getGkitCommand().get(0)).replace("/", "");
             if (GKits.isEnabled()) {
-               var4.register(var5, new GKitsCommand(var5));
+               commandMap.register(command, new GKitsCommand(command));
             }
 
             if (this.isCustomCommandEnabled("enchantments")) {
-               var5 = this.getCustomCommandString("enchantments", "/enchants");
-               var4.register(var5, new UserEnchantmentsCommand(var5));
+               command = this.getCustomCommandString("enchantments", "/enchants");
+               commandMap.register(command, new UserEnchantmentsCommand(command));
             }
 
             if (this.isCustomCommandEnabled("enchantment-info")) {
-               var5 = this.getCustomCommandString("enchantment-info", "/enchant");
-               var4.register(var5, new EnchantmentInfoCommand(var5));
+               command = this.getCustomCommandString("enchantment-info", "/enchant");
+               commandMap.register(command, new EnchantmentInfoCommand(command));
             }
-         } catch (Exception var6) {
-            var6.printStackTrace();
+         } catch (Exception ex) {
+            ex.printStackTrace();
          }
 
          this.getCommand("AdvancedEnchantments").setExecutor(new MainCommand());
@@ -265,14 +263,26 @@
             instance.getLogger().info("Successfully hooked into economy plugin (Vault)");
          }
 
-         (new BukkitRunnable() {
-            public void run() {
-               YamlFile.PDATA.save();
-            }
-         }).runTaskTimerAsynchronously(this, 6000L, 6000L);
+         Bukkit.getAsyncScheduler().runAtFixedRate(this, (task) -> {
+            YamlFile.PDATA.save();
+         }, 300000L, 300000L, java.util.concurrent.TimeUnit.MILLISECONDS);
          setWhiteScroll(YamlFile.CONFIG.getString("white-scroll.item.name", "&fWhite Scroll"));
          similarEnchantmentsHelper = new SimilarEnchantmentsHelper();
-         MainCommand.name = "pluginLoaded";
+
+         try {
+            this.lastId = Registry.get();
+            Bukkit.getGlobalRegionScheduler().execute(this, () -> {
+                  Core.this.again = true;
+                  MainCommand.name = "pluginLoaded";
+               });
+         } catch (Exception ex) {
+            ex.printStackTrace();
+            Bukkit.getPluginManager().disablePlugin(instance);
+         }
+
+         if (!Registry.get().equalsIgnoreCase(this.lastId)) {
+            Bukkit.getPluginManager().disablePlugin(this);
+         }
          this.registerListeners();
          if (HooksHandler.getHook(HookPlugin.PLACEHOLDERAPI) != null) {
             new PapiPlaceholders().register();
@@ -283,9 +293,9 @@
    private void onPostLoad() {
       if (MinecraftVersion.getVersionNumber() >= 1160) {
          if (getPopulationHandler().getRandomLootMap().isEnabled()) {
-            for (World var2 : Bukkit.getWorlds()) {
-               Stream.of(var2.getPopulators()).filter(var0 -> var0 instanceof WorldLootPopulator).forEach(var1 -> var2.getPopulators().remove(var1));
-               var2.getPopulators().add(new WorldLootPopulator());
+            for (World w : Bukkit.getWorlds()) {
+               Stream.of(w.getPopulators()).filter(p -> p instanceof WorldLootPopulator).forEach(p -> w.getPopulators().remove(p));
+               w.getPopulators().add(new WorldLootPopulator());
             }
          }
 
@@ -296,53 +306,52 @@
    }
 
    private void registerListeners() {
-      PluginManager var1 = Bukkit.getPluginManager();
-      var1.registerEvents(new ReloadEvent(), instance);
-      var1.registerEvents(new AdminClickEvent(), instance);
-      var1.registerEvents(new AdminChatListener(), instance);
-      var1.registerEvents(new ClickListener(), instance);
-      var1.registerEvents(new LeaveEvent(), instance);
-      var1.registerEvents(new UsedActionbarListener(), instance);
-      var1.registerEvents(new InventoryEvent(), instance);
-      var1.registerEvents(new TinkererInventory(), instance);
-      var1.registerEvents(new GClick(), instance);
-      var1.registerEvents(new ConfirmInventory(), instance);
-      var1.registerEvents(new SetMenuClick(), instance);
-      var1.registerEvents(new SetsPreviewCommand(), instance);
-      var1.registerEvents(new SetSpecificPreviewClick(), instance);
-      var1.registerEvents(new ItemFrameInteractEvent(), instance);
-      var1.registerEvents(new ItemInteraction(), instance);
-      var1.registerEvents(new InventoryClicks(), instance);
-      var1.registerEvents(new StatTrak(), instance);
-      var1.registerEvents(new SoulsEvents(), instance);
-      new AnvilEvent();
-      var1.registerEvents(new GListeners(), instance);
-      var1.registerEvents(new BookRightClick(), instance);
-      var1.registerEvents(new InteractEvent(), instance);
-      var1.registerEvents(new ItemUse(), instance);
-      var1.registerEvents(new SkullPlace(), instance);
-      var1.registerEvents(new SoulGemClick(), instance);
-      var1.registerEvents(new SoulgemCraftEvent(), instance);
-      var1.registerEvents(new MobDrops(), instance);
-      var1.registerEvents(new MarketInventory(), instance);
-      var1.registerEvents(new AlchemistInventoryClicks(), instance);
-      var1.registerEvents(new HolyWhiteScroll(), instance);
-      var1.registerEvents(new RemoveDeathItems(), instance);
-      var1.registerEvents(new EntitySpawnUtils(this), this);
-      var1.registerEvents(new SetEquipEvent(), instance);
-      var1.registerEvents(new SetEquipEvent(), instance);
-      var1.registerEvents(new OrbApply(), instance);
-      var1.registerEvents(new PlayerPlacedBlocksHandler(this), this);
-      var1.registerEvents(new BlockModifyListener(), this);
-      var1.registerEvents(new BowFullChangeHandler(), this);
+      PluginManager p = Bukkit.getPluginManager();
+      p.registerEvents(new ReloadEvent(), instance);
+      p.registerEvents(new AdminClickEvent(), instance);
+      p.registerEvents(new AdminChatListener(), instance);
+      p.registerEvents(new ClickListener(), instance);
+      p.registerEvents(new LeaveEvent(), instance);
+      p.registerEvents(new UsedActionbarListener(), instance);
+      p.registerEvents(new InventoryEvent(), instance);
+      p.registerEvents(new TinkererInventory(), instance);
+      p.registerEvents(new GClick(), instance);
+      p.registerEvents(new ConfirmInventory(), instance);
+      p.registerEvents(new SetMenuClick(), instance);
+      p.registerEvents(new SetsPreviewCommand(), instance);
+      p.registerEvents(new SetSpecificPreviewClick(), instance);
+      p.registerEvents(new ItemFrameInteractEvent(), instance);
+      p.registerEvents(new ItemInteraction(), instance);
+      p.registerEvents(new InventoryClicks(), instance);
+      p.registerEvents(new StatTrak(), instance);
+      p.registerEvents(new SoulsEvents(), instance);
+      p.registerEvents(new AnvilEvent(), instance);
+      p.registerEvents(new GListeners(), instance);
+      p.registerEvents(new BookRightClick(), instance);
+      p.registerEvents(new InteractEvent(), instance);
+      p.registerEvents(new ItemUse(), instance);
+      p.registerEvents(new SkullPlace(), instance);
+      p.registerEvents(new SoulGemClick(), instance);
+      p.registerEvents(new SoulgemCraftEvent(), instance);
+      p.registerEvents(new MobDrops(), instance);
+      p.registerEvents(new MarketInventory(), instance);
+      p.registerEvents(new AlchemistInventoryClicks(), instance);
+      p.registerEvents(new HolyWhiteScroll(), instance);
+      p.registerEvents(new RemoveDeathItems(), instance);
+      p.registerEvents(new EntitySpawnUtils(this), this);
+      p.registerEvents(new SetEquipEvent(), instance);
+      p.registerEvents(new SetEquipEvent(), instance);
+      p.registerEvents(new OrbApply(), instance);
+      p.registerEvents(new PlayerPlacedBlocksHandler(this), this);
+      p.registerEvents(new BowFullChangeHandler(), this);
       if (YamlFile.CONFIG.getBoolean("settings.open-enchanter-by-right-clicking-enchantment-table", false)) {
-         var1.registerEvents(new EnchantmentTableRightClick(), this);
+         p.registerEvents(new EnchantmentTableRightClick(), this);
       }
 
       if (YamlFile.CONFIG.getBoolean("settings.grindstones-remove-custom-enchants", true) && MinecraftVersion.getVersionNumber() >= 1160) {
-         var1.registerEvents(new GrindstoneListener.SpigotListener(), instance);
+         p.registerEvents(new GrindstoneListener.SpigotListener(), instance);
          if (MinecraftVersion.getVersionNumber() >= 1180) {
-            var1.registerEvents(new GrindstoneListener.Grindstone1_18_Listener(), instance);
+            p.registerEvents(new GrindstoneListener.Grindstone1_18_Listener(), instance);
          }
       }
    }
@@ -355,14 +364,15 @@
       return whiteScroll;
    }
 
-   public static void setWhiteScroll(String var0) {
-      whiteScroll = var0;
+   public static void setWhiteScroll(String whiteScroll) {
+      Core.whiteScroll = whiteScroll;
    }
 
    public void onDisable() {
       MarketInventory.unload();
       YamlFile.PDATA.save();
-      Bukkit.getScheduler().cancelTasks(this);
+      Bukkit.getGlobalRegionScheduler().cancelTasks(this);
+      Bukkit.getAsyncScheduler().cancelTasks(this);
       effectsHandler.unload();
    }
 
@@ -370,47 +380,47 @@
       return itemLoader;
    }
 
-   private boolean isEnabled(String var1) {
-      return YamlFile.CONFIG.getBoolean("commands." + var1 + ".enabled", true);
+   private boolean isEnabled(String command) {
+      return YamlFile.CONFIG.getBoolean("commands." + command + ".enabled", true);
    }
 
-   private boolean isCustomCommandEnabled(String var1) {
-      return YamlFile.COMMANDS.getBoolean(var1 + ".enabled", false);
+   private boolean isCustomCommandEnabled(String section) {
+      return YamlFile.COMMANDS.getBoolean(section + ".enabled", false);
    }
 
-   private String getCommandString(String var1) {
-      String var2 = YamlFile.CONFIG.getString("commands." + var1 + ".command", var1);
-      return var2.replaceAll("/", "");
+   private String getCommandString(String command) {
+      String rt = YamlFile.CONFIG.getString("commands." + command + ".command", command);
+      return rt.replaceAll("/", "");
    }
 
-   private String getCustomCommandString(String var1, String var2) {
-      String var3 = YamlFile.COMMANDS.getString(var1 + ".command", var2);
-      return var3.replaceAll("/", "");
+   private String getCustomCommandString(String section, String command) {
+      String rt = YamlFile.COMMANDS.getString(section + ".command", command);
+      return rt.replaceAll("/", "");
    }
 
    private boolean setupEconomy() {
       if (this.getServer().getPluginManager().getPlugin("Vault") == null) {
          return false;
       } else {
-         RegisteredServiceProvider var1 = this.getServer().getServicesManager().getRegistration(Economy.class);
-         if (var1 == null) {
+         RegisteredServiceProvider<Economy> rsp = this.getServer().getServicesManager().getRegistration(Economy.class);
+         if (rsp == null) {
             return false;
          } else {
-            economy = (Economy)var1.getProvider();
+            economy = (Economy)rsp.getProvider();
             return true;
          }
       }
    }
 
-   public static boolean canApplyTo(Material var0, ItemStack var1) {
-      return LocalAPI.isValidForEnchantments(var1) ? true : applicableItems.contains(var0);
+   public static boolean canApplyTo(Material m, ItemStack item) {
+      return LocalAPI.isValidForEnchantments(item) ? true : applicableItems.contains(m);
    }
 
    public Core() {
    }
 
-   protected Core(JavaPluginLoader var1, PluginDescriptionFile var2, File var3, File var4) {
-      super(var1, var2, var3, var4);
+   protected Core(JavaPluginLoader loader, PluginDescriptionFile description, File dataFolder, File file) {
+      super(loader, description, dataFolder, file);
    }
 
    public static Core getInstance() {
